src/parser/parser.rs

```diff
@@ -123,81 +143,98 @@ fn parse_chain_in_operand(rule: Pair<Rule>) -> Operand {
 }
 
-fn parse_logic(mut pairs: Pairs<Rule>) -> FilterExpression {
+fn parse_logic(pairs: Pairs<Rule>) -> Result<FilterExpression, JsonPathParserError> {
     let mut expr: Option<FilterExpression> = None;
-    while pairs.peek().is_some() {
-        let next_expr = parse_logic_and(pairs.next().unwrap().into_inner());
+    let error_message = format!("Failed to parse logical expression: {:?}", pairs);
+    for pair in pairs {
+        let next_expr = parse_logic_and(pair.into_inner())?;
         match expr {
             None => expr = Some(next_expr),
             Some(e) => expr = Some(Or(Box::new(e), Box::new(next_expr))),
         }
     }
-    expr.unwrap()
+    match expr {
+        Some(expr) => Ok(expr),
+        None => Err(JsonPathParserError::ParserError(error_message)),
+    }
 }
 
-fn parse_logic_and(mut pairs: Pairs<Rule>) -> FilterExpression {
+fn parse_logic_and(pairs: Pairs<Rule>) -> Result<FilterExpression, JsonPathParserError> {
     let mut expr: Option<FilterExpression> = None;
-
-    while pairs.peek().is_some() {
-        let next_expr = parse_logic_atom(pairs.next().unwrap().into_inner());
+    let error_message = format!("Failed to parse logical `and` expression: {:?}", pairs,);
+    for pair in pairs {
+        let next_expr = parse_logic_atom(pair.into_inner())?;
         match expr {
             None => expr = Some(next_expr),
             Some(e) => expr = Some(And(Box::new(e), Box::new(next_expr))),
         }
     }
-    expr.unwrap()
+    match expr {
+        Some(expr) => Ok(expr),
+        None => Err(JsonPathParserError::ParserError(error_message)),
+    }
 }
 
-fn parse_logic_atom(mut pairs: Pairs<Rule>) -> FilterExpression {
-    match pairs.peek().map(|x| x.as_rule()) {
-        Some(Rule::logic) => parse_logic(pairs.next().unwrap().into_inner()),
-        Some(Rule::atom) => {
-            let left: Operand = parse_atom(pairs.next().unwrap());
-            if pairs.peek().is_none() {
-                FilterExpression::exists(left)
-            } else {
-                let sign: FilterSign = FilterSign::new(pairs.next().unwrap().as_str());
-                let right: Operand = parse_atom(pairs.next().unwrap());
-                FilterExpression::Atom(left, sign, right)
+fn parse_logic_atom(mut pairs: Pairs<Rule>) -> Result<FilterExpression, JsonPathParserError> {
+    if let Some(rule) = pairs.peek().map(|x| x.as_rule()) {
+        match rule {
+            Rule::logic => parse_logic(pairs.next().unwrap().into_inner()),
+            Rule::atom => {
+                let left: Operand = parse_atom(pairs.next().unwrap())?;
+                if pairs.peek().is_none() {
+                    Ok(FilterExpression::exists(left))
+                } else {
+                    let sign: FilterSign = FilterSign::new(pairs.next().unwrap().as_str());
+                    let right: Operand = parse_atom(pairs.next().unwrap())?;
+                    Ok(FilterExpression::Atom(left, sign, right))
+                }
```

Comment:
We're still using `.unwrap()` here but it should be sage as we're peeking ahead to make sure there's a value.